x/lscosmos/keeper/delegation_strategy.go

```diff
@@ -300,3 +300,123 @@ func (k Keeper) GetAllValidatorsState(ctx sdk.Context, denom string) (types.Allo
+	// return nil if the new amount to be delegated is already zero
+	if newAmount.IsZero() {
+		return nil, nil
+	}
+
+	// get validators weight map
+	validatorWeightsMap := types.GetValidatorWeightsMap(valList)
+
+	// get total delegations post adding more to it
+	totalDelegationAfterAddingNewAmount := delegationState.TotalDelegations(newAmount.Denom).Add(newAmount).Amount
+
+	// get the difference, current and ideally delegated amounts for each validator
+	differenceCurrentAndIdealAmounts := GetDifferenceCurrentAndIdealAmounts(valList, delegationState, totalDelegationAfterAddingNewAmount)
+
+	// sort the difference, current and ideally delegated amounts for each validator based
+	// on the difference in descending order
+	sort.SliceStable(
+		differenceCurrentAndIdealAmounts,
+		func(i, j int) bool {
+			return differenceCurrentAndIdealAmounts[i].Diff.GT(differenceCurrentAndIdealAmounts[j].Diff)
+		},
+	)
+
+	// this is the first round of distribution to all the validators who are not
+	// yet matched to the ideal distribution
+	finalDistribution := make(map[string]sdk.Coin)
+	for _, i := range differenceCurrentAndIdealAmounts {
+
+		if i.Diff.GT(sdk.ZeroDec()) {
+			if newAmount.Amount.ToDec().LT(i.Diff.Abs()) && newAmount.Amount.Sub(newAmount.Amount).GTE(sdk.OneInt()) {
+				finalDistribution[i.Address] = newAmount
+				newAmount.Amount = newAmount.Amount.Sub(newAmount.Amount)
+			} else if newAmount.Amount.ToDec().GTE(i.Diff.Abs()) && newAmount.Amount.Sub(newAmount.Amount).GTE(sdk.OneInt()) {
+				finalDistribution[i.Address] = sdk.NewCoin(newAmount.Denom, i.Diff.Abs().TruncateInt())
+				newAmount.Amount = newAmount.Amount.Sub(i.Diff.Abs().TruncateInt())
+			}
+		}
+	}
+
+	// this is the second round of distribution to all the validators based on their
+	// current weights
+	temporaryCoin := sdk.NewCoin(newAmount.Denom, sdk.ZeroInt())
+	for _, i := range differenceCurrentAndIdealAmounts {
+		if newAmount.IsZero() {
+			break
+		}
+		amountForValidator := newAmount.Amount.ToDec().Mul(validatorWeightsMap[i.Address]).TruncateInt()
+		if !amountForValidator.IsZero() {
+			finalDistribution[i.Address] = sdk.NewCoin(newAmount.Denom, amountForValidator)
+			temporaryCoin.Amount = temporaryCoin.Amount.Add(amountForValidator)
+		}
+	}
+
+	// convert distribution map into types.ValAddressAmounts array
+	var finalValAddressAmounts types.ValAddressAmounts
+	for key, value := range finalDistribution {
+		finalValAddressAmounts = append(finalValAddressAmounts, types.ValAddressAmount{
+			ValidatorAddr: key,
+			Amount:        value,
+		})
+	}
+
+	// sort the types.ValAddressAmounts based on the validator addresses
+	sort.SliceStable(finalValAddressAmounts, func(i, j int) bool {
+		return finalValAddressAmounts[i].ValidatorAddr < finalValAddressAmounts[j].ValidatorAddr
+	})
+
+	// this is a boundary case when there is still some residue left after
```

Comment:
in this case, if you are starting with 0, and there is 10K delegations, will it mean all the 10K go to single validator?